### DISCUSSION TITLE
Template variable error: 'jinja2.utils.Namespace object' has no attribute 'source'

### DISCUSSION THREAD

**Discussion by hauca** [HASwitchPlate/HASPone#71] [Q&A]
I was able to get my first HASP up and running without any issues. Frankly I was amazed at simplicity to get it working! However as I was adding more items to the HASP I started having problems. I don't know for sure but issues seemed to start when I forgot to add the device name before I saved. I went back added the device and saved but afterwards nothing seems to be working. 

I tried to setup a complete new device and name but still not able to get it working. I am seeing issues in the HA logs about an object missing. I tried the blueprints pre and post the "Break Fix". I'm running current firmware that was built locally and the April release of HA. 

Here is the error I'm seeing in the logs

`Logger: homeassistant.helpers.template
Source: helpers/template.py:1346
First occurred: 6:11:10 PM (1 occurrences)
Last logged: 6:11:10 PM

Template variable error: 'jinja2.utils.Namespace object' has no attribute 'source' when rendering '{%- set color = namespace() -%} {%- for entity in device_entities(haspdevice) -%} {%- if entity|regex_search("^light\..*_selected_foreground_color(?:_\d+|)$") -%} {%- set color.source=entity -%} {{ break }} {%- endif -%} {%- endfor -%} {%- set brightness = state_attr(color.source, "brightness")|int / 255 -%} {%- set red=(state_attr(color.source, "rgb_color")[0] * brightness)|int -%} {%- set green=(state_attr(color.source, "rgb_color")[1] * brightness)|int -%} {%- set blue=(state_attr(color.source, "rgb_color")[2] * brightness)|int -%} {%- set colorcode = (red|bitwise_and(248)*256) + (green|bitwise_and(252)*8) + (blue|bitwise_and(248)/8)|int -%} [{%- if not page_scroll -%}{%- for p in range(1,12) %}{%- if p == page1page|int %}"p[{{p}}].b[1].pco={{colorcode}}"{%- else -%}"p[{{p}}].b[1].pco2={{colorcode}}"{%- endif -%},{%- endfor -%} {%- else -%}{%- for p in range(1,12) %}"p[{{p}}].b[1].pco2={{colorcode}}",{%- endfor -%}{%- endif -%} {%- for p in range(1,12) %}{%- if p == page2page|int %}"p[{{p}}].b[2].pco={{colorcode}}"{%- else -%}"p[{{p}}].b[2].pco2={{colorcode}}"{%- endif -%},{%- endfor -%} {%- if not page_scroll -%}{%- for p in range(1,12) %}{%- if p == page3page|int %}"p[{{p}}].b[3].pco={{colorcode}}"{%- else -%}"p[{{p}}].b[3].pco2={{colorcode}}"{%- endif -%}{% if not loop.last %},{% endif %}{%- endfor -%} {%- else -%}{%- for p in range(1,12) %}"p[{{p}}].b[3].pco2={{colorcode}}"{% if not loop.last %},{% endif %}{%- endfor -%}{%- endif -%}]'`

**Comment by aderusha** [comment]
This turns out to be the result of [a breaking change in 2021.04](https://github.com/home-assistant/core/pull/48140).  I'm still trying to figure out how to address this and have reached out for some help [here](https://community.home-assistant.io/t/2021-04-breaking-changes-and-catching-run-actions/297977).

**Reply by aderusha** [reply]
Allright, I've pushed out updates to the blueprints to address this breaking change.  You'll have to delete and re-import all blueprints to fix this issue :(

**Reply by jhhernan** [reply]
Excuse me, What happens with the automations created with the old bluprints? They continue working normally?

**Reply by aderusha** [reply]
Yup!  If you delete the underlying blueprint, just be sure to re-import the same blueprint and automations created will carry over.

**Comment by jhhernan** [comment]
Hi @aderusha I'm receiving this in my log. I'm not sure if it's related with a breaking change in 2021.04

`2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^sensor\.") -%}
    {{- entity|regex_replace(find="^sensor\.", replace="", ignorecase=true)|regex_replace(find="_sensor(?:_\d+|)$", replace="", ignorecase=true) -}}
    {{ break }}
  {%- endif -%}
{%- endfor -%}'
2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^sensor\..+_sensor(?:_\d+|)$") -%}
    {{ entity }}
    {{ break }}
  {%- endif -%}
{%- endfor -%}'
2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- set color = namespace() -%} {%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^light\..*_selected_foreground_color(?:_\d+|)$") -%}
    {%- set color.source=entity -%}
    {{ break }}
  {%- endif -%}
{%- endfor -%} {%- set brightness = state_attr(color.source, "brightness")|int / 255 -%} {%- set red=(state_attr(color.source, "rgb_color")[0] * brightness)|int -%} {%- set green=(state_attr(color.source, "rgb_color")[1] * brightness)|int -%} {%- set blue=(state_attr(color.source, "rgb_color")[2] * brightness)|int -%} {{ (red|bitwise_and(248)*256) + (green|bitwise_and(252)*8) + (blue|bitwise_and(248)/8)|int }}'
2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- set color = namespace() -%} {%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^light\..*_selected_background_color(?:_\d+|)$") -%}
    {%- set color.source=entity -%}
    {{ break }}
  {%- endif -%}
{%- endfor -%} {%- set brightness = state_attr(color.source, "brightness")|int / 255 -%} {%- set red=(state_attr(color.source, "rgb_color")[0] * brightness)|int -%} {%- set green=(state_attr(color.source, "rgb_color")[1] * brightness)|int -%} {%- set blue=(state_attr(color.source, "rgb_color")[2] * brightness)|int -%} {{ (red|bitwise_and(248)*256) + (green|bitwise_and(252)*8) + (blue|bitwise_and(248)/8)|int }}'
2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- set color = namespace() -%} {%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^light\..*_unselected_foreground_color(?:_\d+|)$") -%}
    {%- set color.source=entity -%}
    {{ break }}
  {%- endif -%}
{%- endfor -%} {%- set brightness = state_attr(color.source, "brightness")|int / 255 -%} {%- set red=(state_attr(color.source, "rgb_color")[0] * brightness)|int -%} {%- set green=(state_attr(color.source, "rgb_color")[1] * brightness)|int -%} {%- set blue=(state_attr(color.source, "rgb_color")[2] * brightness)|int -%} {{ (red|bitwise_and(248)*256) + (green|bitwise_and(252)*8) + (blue|bitwise_and(248)/8)|int }}'
2021-04-10 18:56:00 WARNING (MainThread) [homeassistant.helpers.template] Template variable warning: 'break' is undefined when rendering '{%- set color = namespace() -%} {%- for entity in device_entities(haspdevice) -%}
  {%- if entity|regex_search("^light\..*_unselected_background_color(?:_\d+|)$") -%}
    {%- set color.source=entity -%}
    {{ break }}
  {%- endif -%}
{%- endfor -%} {%- set brightness = state_attr(color.source, "brightness")|int / 255 -%} {%- set red=(state_attr(color.source, "rgb_color")[0] * brightness)|int -%} {%- set green=(state_attr(color.source, "rgb_color")[1] * brightness)|int -%} {%- set blue=(state_attr(color.source, "rgb_color")[2] * brightness)|int -%} {{ (red|bitwise_and(248)*256) + (green|bitwise_and(252)*8) + (blue|bitwise_and(248)/8)|int }}'
`

**Reply by aderusha** [reply]
Delete and re-import the blueprint and you should be good to go!

**Reply by jhhernan** [reply]
Thanks but still getting the same message

**Reply by aderusha** [reply]
If you're still seeing the message, the re-import somehow didn't work.  Be sure to delete all existing blueprints first, then re-import.  The error above that you are seeing is 100% due to the old blueprints still being used somehow, so one way or another we're going to have to get rid of those.

Let me know how it goes, and if it's still not working we can try walking through some additional steps to make certain the old blueprints are gone.

**Reply by jhhernan** [reply]
Thanks @aderusha . Your are right!  I repeat the process. I delete all the bluepints, reboot my machine and then re-import the bueprints. It appears the warnings are gone. Thank you!

**Comment by damanti-me** [comment]
https://github.com/HASwitchPlate/HASPone/issues/67
